File: 31-next-permutation/31-next-permutation.go
func nextPermutation(nums []int)  {
  n := len(nums)
  k := n-2
  for ; k >= 0; k-- {
    if nums[k] < nums[k+1] {
      break
    }
  }
  if k < 0 {
    for i, j := 0, n-1; i < j; i, j = i+1, j-1 {
      nums[i], nums[j] = nums[j], nums[i]
    }
    return 
  }
  v := nums[k]
  l := n-1
  for ; l > k; l-- {
    if nums[l] > v {
      break
    }
  }
  nums[k], nums[l] = nums[l], v
  for i, j := k+1, n-1; i < j; i, j = i+1, j-1 {
    nums[i], nums[j] = nums[j], nums[i]
  }
}

File: 31-next-permutation/31-next-permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        n = len(nums);
        k = n - 2
        while k >= 0 and nums[k] >= nums[k+1]:
          k = k - 1
        if k < 0:
          nums[:] = nums[::-1]
        else:
          v = nums[k]
          l = n - 1
          while l > k and nums[l] <= v:
            l = l - 1
          nums[l], nums[k] = v, nums[l]
          nums[k+1:] = nums[k+1:][::-1]

        